File: sdk/Managed/src/Microsoft.WindowsAzure.MobileServices/Table/Query/OData/BinaryOperatorNode.cs
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Microsoft.WindowsAzure.MobileServices.Query
{
    /// <summary>
    /// Query node representing a binary operator.
    /// </summary>
    public sealed class BinaryOperatorNode: QueryNode
    {
        /// <summary>
        /// Gets the kind of the query node.
        /// </summary>
        public override QueryNodeKind Kind
        {
            get { return QueryNodeKind.BinaryOperator; }
        }

        /// <summary>
        /// The operator represented by this node.
        /// </summary>
        public BinaryOperatorKind OperatorKind { get; private set; }

        /// <summary>
        /// Gets the left operand.
        /// </summary>
        public QueryNode LeftOperand { get; internal set; }

        /// <summary>
        /// Gets the right operand.
        /// </summary>
        public QueryNode RightOperand { get; internal set; }

        /// <summary>
        /// Initializes instance of <see cref="BinaryOperatorNode"/>
        /// </summary>
        /// <param name="kind">The kind of this node.</param>
        /// <param name="left">The left operand.</param>
        /// <param name="right">The right operand.</param>
        public BinaryOperatorNode(BinaryOperatorKind kind, QueryNode left, QueryNode right)
        {
            this.OperatorKind = kind;
            this.LeftOperand = left;
            this.RightOperand = right;
        }

        /// <summary>
        /// Accept a <see cref="QueryNodeVisitor{T}" /> to walk a tree of <see cref="QueryNode" />s.
        /// </summary>
        /// <typeparam name="T">Type that the visitor will return after visiting this token.</typeparam>
        /// <param name="visitor">An implementation of the visitor interface.</param>
        /// <returns>An object whose type is determined by the type parameter of the visitor.</returns>
        public override T Accept<T>(QueryNodeVisitor<T> visitor)
        {
            return visitor.Visit(this);
        }        

        internal override void SetChildren(IList<QueryNode> children)
        {
            Debug.Assert(children.Count == 2);

            this.LeftOperand = children[0];
            this.RightOperand = children[1];
        }
    }
}

File: sdk/android/src/sdk/src/main/java/com/microsoft/windowsazure/mobileservices/table/query/BinaryOperatorNode.java
/*
Copyright (c) Microsoft Open Technologies, Inc.
All Rights Reserved
Apache 2.0 License
 
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
 
     http://www.apache.org/licenses/LICENSE-2.0
 
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 
See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */

/**
 * BinaryOperatorNode.java
 */
package com.microsoft.windowsazure.mobileservices.table.query;

/**
 * Class that represents a binary operator query node
 */
class BinaryOperatorNode implements QueryNode {
    private BinaryOperatorKind mBinaryOperatorKind;
    private QueryNode mLeftArgument;
    private QueryNode mRightArgument;

    /**
     * Constructor for BinaryOperatorNode
     *
     * @param binaryOperatorKind The binary operator kind
     */
    BinaryOperatorNode(BinaryOperatorKind binaryOperatorKind) {
        this.mBinaryOperatorKind = binaryOperatorKind;
    }

    @Override
    public QueryNode deepClone() {
        BinaryOperatorNode clone = new BinaryOperatorNode(this.mBinaryOperatorKind);

        clone.mLeftArgument = this.mLeftArgument.deepClone();
        clone.mRightArgument = this.mRightArgument.deepClone();

        return clone;
    }

    @Override
    public QueryNodeKind getKind() {
        return QueryNodeKind.BinaryOperator;
    }

    @Override
    public <T> T accept(QueryNodeVisitor<T> visitor) {
        return visitor.visit(this);
    }

    /**
     * Gets the kind of binary operator node.
     */
    BinaryOperatorKind getBinaryOperatorKind() {
        return this.mBinaryOperatorKind;
    }

    /**
     * Gets the left argument query node of the binary operator node.
     */
    QueryNode getLeftArgument() {
        return this.mLeftArgument;
    }

    /**
     * Sets the left argument query node of the binary operator node.
     */
    void setLeftArgument(QueryNode leftArgument) {
        this.mLeftArgument = leftArgument;
    }

    /**
     * Gets the right argument query node of the binary operator node.
     */
    QueryNode getRightArgument() {
        return this.mRightArgument;
    }

    /**
     * Sets the right argument query node of the binary operator node.
     */
    void setRightArgument(QueryNode rightArgument) {
        this.mRightArgument = rightArgument;
    }
}